### PR TITLE
MAGICDRAW-868

### PR DESCRIPTION
**Pull request by w3e**
SR now first creates the list of all redefinable elements and filters out the not-leafes.
Fixes issue with circular reference warning.

### CODE REVIEW

File: src/main/java/gov/nasa/jpl/mbee/mdk/systems_reasoner/actions/SpecializeStructureAction.java [hunk 1]
```diff
@@ -113,8 +113,12 @@ public Classifier createSpecialClassifier(Namespace container, List<RedefinableE
         }
         Utils.createGeneralization(classifier, specific);
 
-
-        for (final NamedElement ne : specific.getInheritedMember()) { // Exclude Classifiers for now -> Should Aspect Blocks be Redefined?
+        Set<NamedElement> listOfAllMembers = new HashSet<>();
```
Comment (ivan-gomes) [inline src/main/java/gov/nasa/jpl/mbee/mdk/systems_reasoner/actions/SpecializeStructureAction.java]: `Set<NamedElement> listOfAllMembers = new HashSet<>(specific.getInheritedMember());` can replace this block.

File: src/main/java/gov/nasa/jpl/mbee/mdk/systems_reasoner/actions/SpecializeStructureAction.java [hunk 2]
```diff
@@ -113,8 +113,12 @@ public Classifier createSpecialClassifier(Namespace container, List<RedefinableE
...
+        for (NamedElement namedElement : specific.getInheritedMember()) {
+            listOfAllMembers.add(namedElement);
+        }
+        filterRedefinedElements(listOfAllMembers);
```
Comment (ivan-gomes) [inline src/main/java/gov/nasa/jpl/mbee/mdk/systems_reasoner/actions/SpecializeStructureAction.java]: No need to separate into a void method if it's only going to be called once.

File: src/main/java/gov/nasa/jpl/mbee/mdk/systems_reasoner/actions/SpecializeStructureAction.java [hunk 3]
```diff
@@ -127,6 +131,27 @@ public Classifier createSpecialClassifier(Namespace container, List<RedefinableE
...
+        List<RedefinableElement> removeElements = new ArrayList<>();
+        for (NamedElement member : listOfAllMembers) {
+            if(member instanceof RedefinableElement){
+                Collection<RedefinableElement> redefinedBy = ((RedefinableElement) member).getRedefinedElement();
+                for (RedefinableElement redefinableElement : redefinedBy) {
+                    removeElements.add((RedefinableElement) redefinableElement);
+                }
+            }
+        }
+        for (RedefinableElement removeElement : removeElements) {
```
Comment (ivan-gomes) [inline src/main/java/gov/nasa/jpl/mbee/mdk/systems_reasoner/actions/SpecializeStructureAction.java]: `listOfAllMembers.removeAll(removeElements);`

File: src/main/java/gov/nasa/jpl/mbee/mdk/systems_reasoner/actions/SpecializeStructureAction.java [hunk 4]
```diff
@@ -127,6 +131,27 @@ public Classifier createSpecialClassifier(Namespace container, List<RedefinableE
...
+    /**
+     * Only leave the lowest element in the redefinition hierarchy.
+     * @param listOfAllMembers
+     */
+    private void filterRedefinedElements(Set<NamedElement> listOfAllMembers) {
+        List<RedefinableElement> removeElements = new ArrayList<>();
+        for (NamedElement member : listOfAllMembers) {
+            if(member instanceof RedefinableElement){
+                Collection<RedefinableElement> redefinedBy = ((RedefinableElement) member).getRedefinedElement();
+                for (RedefinableElement redefinableElement : redefinedBy) {
```
Comment (ivan-gomes) [inline src/main/java/gov/nasa/jpl/mbee/mdk/systems_reasoner/actions/SpecializeStructureAction.java]: `removeElements.addAll(redefinableElement);`

File: src/main/java/gov/nasa/jpl/mbee/mdk/systems_reasoner/actions/CreateSpecializedTypeAction.java [hunk 5]
```diff
@@ -48,7 +48,7 @@ public static final boolean createSpecializedType(final StructuralFeature redefi
             Application.getInstance().getGUILog().log(parent.getQualifiedName() + " is not editable. Skipping creating specialization.");
             return true;
         }
-        if (redefinedAttribute.getType() instanceof Classifier && !(redefinedAttribute.getType() instanceof Property)) {
+        if (!(redefinedAttribute.getType() instanceof DataType) && !(redefinedAttribute.getType() instanceof Property)) {
```
Comment (ivan-gomes) [inline src/main/java/gov/nasa/jpl/mbee/mdk/systems_reasoner/actions/CreateSpecializedTypeAction.java]: Why did we make this change?

Comment (w3e) [reply src/main/java/gov/nasa/jpl/mbee/mdk/systems_reasoner/actions/CreateSpecializedTypeAction.java]: Realized that DataType evaluates to true under instanceof Classifier test.

**Comment by w3e** [comment]
Added changes.

**Comment by ivan-gomes** [comment]
This pull request fixes 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/Open-MBEE/MDK/rev/pr-424fb38d70d550b7cee2652d62d187d45825dcdb)

**fixed alerts:**

* 1 for Container contents are never accessed

---

*Comment posted by [lgtm.com](https://lgtm.com)*

**Comment by w3e** [comment]
Looks good now, lgtm issue seemed to be older.

**Comment by ivan-gomes** [comment]
This pull request fixes 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/Open-MBEE/MDK/rev/pr-ab5e7a5c7c38ed3aba812b0176873dc12c721eaa)

**fixed alerts:**

* 1 for Container contents are never accessed

---

*Comment posted by [lgtm.com](https://lgtm.com)*

**Comment by ivan-gomes** [comment]
This pull request fixes 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/Open-MBEE/MDK/rev/pr-ef92f7fe916cff7d0beef167f98eb556713ec220)

**fixed alerts:**

* 1 for Container contents are never accessed

---

*Comment posted by [lgtm.com](https://lgtm.com)*

**Pull request merged**